### PR TITLE
apply traffic shaping on subgraph directly

### PR DESCRIPTION
**Pull request by Geal**
The plugin infrastructure works on `BoxService` instances, and makes no guarantee on plugin ordering.
The traffic shaping plugin needs a clonable inner service, and should run right before calling
the underlying service. So this changes the traffic plugin application so it can work directly
on the underlying service. It is still a plugin though, so it keeps the same configuration.

### CODE REVIEW

**Comment by BrynCooke** [comment]
Is there a way we could extend the existing functionality around plugins ordering to allow traffic shaping to be moved to last?
https://github.com/apollographql/router/blob/dev/apollo-router/src/router_factory.rs#L228

If this would work then I think it'd be cleaner.

**Comment by Geal** [comment]
we could, but the service it manipulates would still be a `BoxService`, which is not clonable. And we need a clonable service for deduplication (that was done manually here before https://github.com/apollographql/router/pull/2034/files#diff-67b9590c5fe973fea43d5bc5a29f37da7b961823665694909f45a00ec50e873dL131 ), retries and caching

File: apollo-router/src/plugins/traffic_shaping/mod.rs [hunk 1]
```diff
@@ -528,17 +604,25 @@ mod test {
         mock_service.expect_clone().returning(|| {
             let mut mock_service = MockSupergraphService::new();
 
-            mock_service.expect_call().times(0..2).returning(move |_| {
-                Ok(SupergraphResponse::fake_builder()
-                    .data(json!({ "test": 1234_u32 }))
-                    .build()
-                    .unwrap())
+            println!("cloning the supergraph");
```
Comment (bnjjj) [inline apollo-router/src/plugins/traffic_shaping/mod.rs:607]: Should we keep it ?

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/services/supergraph_service.rs [hunk 2]
```diff
@@ -408,23 +410,35 @@ impl RouterCreator {
...
+                plugins: self.plugins.clone(),
+                subgraph_creator: self.subgraph_creator.clone(),
+            })
+            .schema(self.schema.clone())
+            .build();
+
+        let supergraph_service = match self
+            .plugins
+            .iter()
+            .find(|i| i.0.as_str() == "apollo.traffic_shaping")
```
Comment (bnjjj) [inline apollo-router/src/services/supergraph_service.rs:426]: As you're using `apollo.traffic_shaping` at different places, could we put this in a constant ? To get rid of typo and so one

Comment (Geal) [reply apollo-router/src/services/supergraph_service.rs]: fixed in https://github.com/apollographql/router/pull/2034/commits/037052927d4401471a1a66400e69f2795863114f

File: apollo-router/src/router_factory.rs [hunk 3]
```diff
@@ -120,17 +118,27 @@ impl SupergraphServiceConfigurator for YamlSupergraphServiceFactory {
     ) -> Result<Self::SupergraphServiceFactory, BoxError> {
         // Process the plugins.
         let plugins = create_plugins(&configuration, &schema, extra_plugins).await?;
-        let deduplicate_queries = is_deduplicate_query_enabled(configuration.as_ref());
 
         let mut builder = PluggableSupergraphServiceBuilder::new(schema.clone());
         builder = builder.with_configuration(configuration);
 
+        println!(
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/router_factory.rs]: removed in https://github.com/apollographql/router/pull/2034/commits/28334c2ac432856dd039996d0926eefef6bc8f54

File: apollo-router/src/plugins/traffic_shaping/mod.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/traffic_shaping/mod.rs]: removed in https://github.com/apollographql/router/pull/2034/commits/28334c2ac432856dd039996d0926eefef6bc8f54

**Pull request merged**